tiledb/sm/cpp_api/ndrectangle.h

```diff
@@ -0,0 +1,239 @@
+/**
+ * @file   ndrectangle.h
+ *
+ * @section LICENSE
+ *
+ * The MIT License
+ *
+ * @copyright Copyright (c) 2024 TileDB, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * This file declares the C++ API for the TileDB NDRectangle.
+ */
+
+#ifndef TILEDB_CPP_API_NDRECTANGLE_H
+#define TILEDB_CPP_API_NDRECTANGLE_H
+
+#include "context.h"
+#include "deleter.h"
+#include "dimension.h"
+#include "domain.h"
+#include "tiledb.h"
+#include "type.h"
+
+#include <functional>
+#include <memory>
+#include <type_traits>
+
+namespace tiledb {
+
+/**
+ * Represents an NDRectangle. This is similar to the non empty domain API with
+ *the exception is doesn’t differentiate between fixed and var size. It will be
+ *used in the future to replace the return value of the non empty domain API.
+ *Note that an NDRectangle object should have all dimensions specified before
+ *it can be used in any call to set an ND rectangle
+ *
+ * @details
+ * **Example:**
+ *
+ * @code{.cpp}
+ *
+ * TODO
+ * @endcode
+ *
+ **/
+class NDRectangle {
+ public:
+  /* ********************************* */
+  /*     CONSTRUCTORS & DESTRUCTORS    */
+  /* ********************************* */
+
+  explicit NDRectangle(const tiledb::Context& ctx, const tiledb::Domain& domain)
+      : ctx_(ctx) {
+    tiledb_ndrectangle_t* capi_ndrect;
+    ctx.handle_error(tiledb_ndrectangle_alloc(
+        ctx.ptr().get(), domain.ptr().get(), &capi_ndrect));
+    ndrect_ = std::shared_ptr<tiledb_ndrectangle_t>(capi_ndrect, deleter_);
+  }
+
+  /**
+   * Loads the NDRectangle of an existing currentDomain with the input C
```

Comment:
The loading part here is confusing, this just wraps a `tiledb_ndrectangle_t*`. Also please specify that the object will manage the lifetime of `tiledb_ndrectangle_t*` arg.